evm/p2p/auth.py

```diff
@@ -0,0 +1,285 @@
+import asyncio
+import logging
+import os
+import random
+import struct
+
+import rlp
+from rlp import sedes
+
+from evm.ecc import get_ecc_backend
+from evm.utils.ecdsa import (
+    decode_signature,
+    encode_signature,
+)
+from evm.utils.keccak import (
+    keccak,
+    keccak_updateable,
+)
+from evm.utils.numeric import (
+    int_to_big_endian,
+)
+from evm.p2p import ecies
+from evm.p2p.constants import (
+    AUTH_ACK_LEN,
+    AUTH_MSG_LEN,
+    ENCRYPTED_AUTH_ACK_LEN,
+    ENCRYPTED_AUTH_MSG_LEN,
+    HASH_LEN,
+    PUBKEY_LEN,
+    SIGNATURE_LEN,
+    SUPPORTED_RLPX_VERSION,
+)
+from evm.p2p.exceptions import AuthenticationError
+from evm.p2p.peer import Peer
+from evm.p2p.utils import (
+    sxor,
+)
+from evm.utils.secp256k1 import (
+    encode_raw_public_key,
+    private_key_to_public_key,
+)
+
+
+@asyncio.coroutine
+def handshake(remote, privkey):
+    """Perform the auth handshake with the given remote and return a Peer ready to be used.
+
+    The Peer will be configured with the shared secrets established during the handshake.
+    """
+    initiator = HandshakeInitiator(remote, privkey)
+    reader, writer = yield from initiator.connect()
+
+    initiator_nonce = keccak(int_to_big_endian(random.randint(0, 2 ** 256 - 1)))
```

Comment:
From #96:
Is it safe for us to be using random.random? My understanding has always been no.

With pycryptodome we can use Crypto.Random.get_random_bytes(n) which I understand to be appropriate for key generation. thoughts?

Reply:
@pipermerriam On py3.6 we could use the new `secrets` module, but I guess we want to support 3.5? Anyway, [the pyca/cryptography docs suggest this alternative](https://cryptography.io/en/latest/random-numbers/): `int.from_bytes(os.urandom(20), byteorder="big")`